activesupport/lib/active_support/deprecation/instance_delegator.rb

```diff
@@ -18,7 +17,30 @@ def include(included_module)
         end
 
         def method_added(method_name)
-          singleton_class.delegate(method_name, to: :instance)
+          use_instead =
+            case method_name
+            when :silence, :behavior=, :disallowed_behavior=, :disallowed_warnings=, :silenced=, :debug=
+              target = "(defined?(Rails.application.deprecators) ? Rails.application.deprecators : ActiveSupport::Deprecation._instance)"
+              "Rails.application.deprecators.#{method_name}"
+            when :warn, :deprecate_methods, :gem_name, :gem_name=, :deprecation_horizon, :deprecation_horizon=
+              "your own Deprecation object"
```

Comment:
Could we give an example here? Consider this code in an app (not in a gem/engine), which I think is pretty common and there is lots of internet recommendations to use:

```ruby
ActiveSupport::Deprecation.warn "foo"
```

You get this warning:

```
DEPRECATION WARNING: Calling warn on ActiveSupport::Deprecation is deprecated and will be removed from Rails (use your own Deprecation object instead)
```

I think a more useful warning would be:

```
DEPRECATION WARNING: Calling warn on ActiveSupport::Deprecation is deprecated and will be removed from Rails (use `ActiveSupport.deprecator.warn` instead)
```

Unless we are recommending that apps define their own deprecators as well? (Which to me seems excessive for most apps.)

Reply:
https://github.com/rails/rails/pull/47978

Reply:
I agree that we could give a better example of how to fix this deprecation. We absolutely do not want to point to `ActiveSupport.deprecator` though. The whole point of this deprecation cycle is to stop third-party code from generating deprecations that seem to come from Rails.

Reply:
But what about first party code? What should it do?

Reply:
I meant third-party from the point of view of Rails, i.e. libraries and the application.

Regarding application deprecations, at Shopify, we've added an application deprecator in `config/application.rb`:

```ruby
module Shopify
  class Application < Rails::Application

    # ...

    initializer :deprecator do |app|
      app.deprecators[:shopify] = ActiveSupport::Deprecation.new("soon", "Shopify/shopify")
    end
  end
end
```

And then we use it directly:
```ruby
Rails.application.deprecators[:shopify].warn("some warning")
```

We've debated whether to do like a gem would typically do and have an API like `Shopify.deprecator` but ultimately felt like the out-of-the-box API from Rails was good enough.

I also thought that maybe Rails could always define an application deprecator to be used directly but didn't propose it because ultimately I don't think most apps need deprecations (it's possible to just change callers and callees at once, unlike in a library/framework), the deprecation horizon doesn't make much sense, etc.

If an application needs a deprecator, it can do what we did above, or do like in a gem, define a singleton method on the application module, set in the application deprecators in an initializer.

I'm not sure what we should write in the deprecation to make that clearer, there's some setup involved… Yes the deprecated API was clean and nice, but in the end it was misguided as separate gems and applications in addition to Rails were all sharing the same `Deprecation` instance.

Reply:
Hmm. I don't think the fact that setup is required should be celebrated. I understand Shopify doing this but for the long tail of small apps I think this is overly complicated vs. what Rails 7 and before offered.